### ISSUE TITLE
Much slower int16->float32 conversion compared to NumPy

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#42959]
I found that int16->float32 conversion in PyTorch is at least 3x slower than in NumPy.

How can we properly measure wall clock time when using multi-threaded code?

`time python3 bug.py`

```
use_numpy False use_torch True set_num_threads False
21875.09536743164 us

real    0m3.086s
user    1m13.246s
sys     0m0.268s
```

```
use_numpy True use_torch False set_num_threads False
18.9971923828125 us

real    0m0.852s
user    0m0.722s
sys     0m0.125s
```

```
use_numpy False use_torch True set_num_threads True
35.28118133544922 us

real    0m1.022s
user    0m0.855s
sys     0m0.154s
```

```python
# bug.py

(use_numpy, use_torch, set_num_threads) = (False, True, False)
print('use_numpy', use_numpy, 'use_torch', use_torch, 'set_num_threads', set_num_threads)

import time
import torch

timer = time.time
s2f = lambda signal: signal.float() / 32767
s2f_ = lambda signal: signal.astype('float32') / 32767

number = 100
scale = 1e-6

signal = torch.randint(-32_000, 32_000, (40_000, ), dtype = torch.int16)

if set_num_threads:
    torch.set_num_threads(1)

if use_torch:
    tic = timer()
    for i in range(number):
        s2f(signal)
    print((timer() - tic) / scale / number, 'us')

if use_numpy:
    tic = timer()
    for i in range(number):
        s2f_(signal.numpy())
    print((timer() - tic) / scale / number, 'us')
```

**Comment by ngimel** [comment]
cc @robieta for perf measurements. I thought time.time() gives wallclock time and not the sum of the threads. `set_num_threads` reliably disables threading, it's not a context manager, so you need to control it manually. Don't mix set_num_threads with OMP_NUM_THREADS/MKL_NUM_THREADS env variables.

**Comment by vadimkantorov** [comment]
> Don't mix set_num_threads with OMP_NUM_THREADS/MKL_NUM_THREADS env variables.

Don't mix in which way? :) If I can get around by only `torch.set_num_threads(...)` of course I'll use only them. Is set_num_threads sufficient? I remember there were a lot of reports over the years that it was not sufficient, but maybe now it's fixed.

**Comment by vadimkantorov** [comment]
> I thought time.time() gives wallclock time and not the sum of the threads.

That's what I thought as well. But real time is different 3x, while report of time.time is different 1000x

**Comment by robieta** [comment]
I think there are a couple things going on here:

<details>

<summary> Benchmark script </summary>

```
import time

import torch
from torch.utils._benchmark import Timer, Compare


create_pt = lambda n: torch.randint(-32_000, 32_000, (n, ), dtype=torch.int16)
create_np = lambda n: create_pt(n).numpy()

to_float_pt = lambda signal: signal.float()
to_float_np = lambda signal: signal.astype('float32')

div_int16max = lambda signal: signal / 32767

def convert_pt(signal):
    return signal.float() / 32767

def convert_np(signal):
    return signal.astype('float32') / 32767

def convert_np_two_pass(signal):
    x = signal.astype('float32')
    return x / 32767


def measure(sub_label, signal, fn):
    timer = Timer(
        stmt="fn(signal)",
        globals={
            "fn": fn,
            "signal": signal,
        },
        description=str(len(signal)),
        label="Int to normalized float",
        sub_label=sub_label,
    )
    return timer.blocked_autorange(min_run_time=3)


results = []
for n in [4_000, 40_000, 400_000, 4_000_000, 40_000_000]:
    results.extend([
        measure("PyTorch: int16 -> float32", create_pt(n), to_float_pt),
        measure("NumPy:   int16 -> float32", create_np(n), to_float_np),

        measure("PyTorch: float32 / 32767", to_float_pt(create_pt(n)), div_int16max),
        measure("NumPy:   float32 / 32767", to_float_np(create_np(n)), div_int16max),

        measure("PyTorch: overall ", create_pt(n), convert_pt),
        measure("NumPy:   overall ", create_np(n), convert_np),
        measure("NumPy:   two pass", create_np(n), convert_np_two_pass),
    ])
    print(f"{n} complete.")

    # Without this statement, numpy two pass is very, very slow.
    time.sleep(1)

cmp = Compare(results)
cmp.trim_significant_figures()
cmp.print()
```

</details>

<details>

<summary> Results </summary>

```
[----------------------------- Int to normalized float ------------------------------]
                                 |  4000  |  40000  |  400000  |  4000000  |  40000000
1 threads: ---------------------------------------------------------------------------
      PyTorch: int16 -> float32  |    6   |    16   |    120   |    1800   |    100000
      NumPy:   int16 -> float32  |    2   |    11   |    110   |    1500   |   4813000

      PyTorch: float32 / 32767   |   11   |    32   |    220   |    2800   |    120000
      NumPy:   float32 / 32767   |    3   |    22   |    200   |    2700   |   4875000

      PyTorch: overall           |   18   |    49   |    320   |    5800   |    230000
      NumPy:   overall           |    5   |    34   |    310   |    3600   |   6147000
      NumPy:   two pass          |    5   |    33   |   1300   |    5800   |  12800000

Times are in microseconds (us).
```

Row separations between the three sections added by me.

</details>

1) The dismal performance in multi-threaded mode is just garden variety contention, which we should fix.
2) For small `n`, the difference is constant(ish) and seems to be purely overhead. Part of that is that NumPy has lower overhead in general. (In fairness, it also has to keep track of less.) The other part seems to be specialized instructions. PyTorch is going through vectorized TensorIterator calls which are reasonably fast, but NumPy has specialized implementations (`_aligned_contig_cast_short_to_float` and `sse2_binary_scalar2_divide_FLOAT`) which seem to be lower overhead, albeit also with a bit worse scaling.
3) NumPy is doing some interesting trace-compiler-esq trick under the hood. Between the `400k` and `4M` cases (which is where you have to spill out of L3 cache) the difference between two almost identical formulations becomes quite significant. (`400k` two pass is sometimes slow and sometimes not. This one it just happened to be. The `4M` result is not a fluke, but is quite repeatable.) I think what's going on is that by creating a Python object which can observe the intermediate array, NumPy is forced to actually materialize the intermediate values. (Which PyTorch does unconditionally.) And by the time you get to `40M`, whatever trick NumPy is doing appears to be past its scalability limit.

As for next steps, while overhead reduction is always useful I don't see anything in these numbers to suggest that is egregious. On the other hand, we definitely need to fix the overzealous multithreading, particularly since it is the default behavior.

**Comment by vadimkantorov** [comment]
A still open question for me:

why the "real" time difference is 3x while time.time difference is 1000x? Does it mean PyTorch should very actively advertise that time.time is not suitable for timing multi-threaded workloads?

**Comment by vadimkantorov** [comment]
In general, it would be good to improve CPU perf (e.g. another recent issue https://github.com/pytorch/pytorch/issues/43033). CPU work is quite frequent, at very least in data loader threads doing exactly these kinds of conversions

**Comment by robieta** [comment]
@ngimel is right: the timer shouldn't cause that kind of deviation. One thing that occurs to me is that your code doesn't have any warmup, so it will measure both the task of interest but also instantiation of framework internals. (e.g. a threadpool) I reran your code with replicates, and I noticed the first run without num_threads set was ~10x slower than subsequent ones. Have you tried `torch.utils._benchmark.Timer`? (Since it handles a lot of these issues out of the box.) When I re-ran my benchmark with `num_threads=24` in `Timer` I got reasonable looking numbers.

I agree that CPU efficiency is quite important; that's actually where I spend the majority of my effort. I was just saying that I didn't see any particularly low hanging fruit in my measurements. One thing to note on the topic of data loaders is that it is often necessary to operate on batches rather than examples if you want peak performance, so as to better amortize overheads.

**Comment by vadimkantorov** [comment]
> One thing to note on the topic of data loaders is that it is often necessary to operate on batches rather than examples if you want peak performance, so as to better amortize overheads.

I agree, but for legacy reasons, many of vision workloads use per-element transforms. Same can be true for audio processing in older systems, where log mel features are computed per-element prior to collating the batch.

**Comment by vadimkantorov** [comment]
```python
# bug.py

(use_numpy, use_torch, set_num_threads) = (False, True, False)
print('use_numpy', use_numpy, 'use_torch', use_torch, 'set_num_threads', set_num_threads)

import time
import torch

timer = time.time
s2f = lambda signal: signal.float() / 32767
s2f_ = lambda signal: signal.astype('float32') / 32767

number = 100
scale = 1e-6

signal = torch.randint(-32_000, 32_000, (40_000, ), dtype = torch.int16)

if set_num_threads:
    torch.set_num_threads(1)

if use_torch:
    for i in range(number):
        s2f(signal)

    tic = timer()
    for i in range(number):
        s2f(signal)
    print((timer() - tic) / scale / number, 'us')

if use_numpy:
    for i in range(number):
        s2f_(signal.numpy())

    tic = timer()
    for i in range(number):
        s2f_(signal.numpy())
    print((timer() - tic) / scale / number, 'us')
```

```
use_numpy False use_torch True set_num_threads False
21213.462352752686 us
use_numpy True use_torch False set_num_threads False
16.145706176757812 us
```
This does warmup, and still shows 1000x difference.

It's good there's now some standard way of benchmarking, but if you look in issues, 70% of benchmarking uses simply time.time() :(

**Comment by vadimkantorov** [comment]
Increasing number to 1000:
```
use_numpy False use_torch True set_num_threads False
70.6014633178711 us

real    0m19.722s
user    10m30.768s
sys     0m0.878s

use_numpy True use_torch False set_num_threads False
15.280008316040039 us

real    0m0.905s
user    0m0.781s
sys     0m0.121s
```

But still, why warmup of 100 times is not sufficient?

**Comment by robieta** [comment]
I'm not able to reproduce your 1000x number, but I do sometimes see spikes. (It seems to be transient, and generally not more than 2x.) Maybe grab some perf stats and see what it's blocking on?

**Comment by ZisIsNotZis** [comment]
Interestingly, I found pytorch to be much faster on this

```python
In [1]: import numpy as np

In [2]: a = np.zeros(2**20, 'H')

In [3]: %timeit a.astype('f')
171 µs ± 336 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [4]: a = np.zeros(2**20, 'h')

In [5]: %timeit a.astype('f')
219 µs ± 770 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [9]: a = torch.zeros(2**20, dtype=torch.int16)

In [6]: import torch

In [10]: %timeit a.float()
81.2 µs ± 310 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

With `numpy=1.19.1` from anaconda built against MKL, `pytorch=1.7.0.dev20200816` from pytorch-nightly.

The 'H' is basically uint16 which pytorch does not support, and 'h' is int16

**Comment by ZisIsNotZis** [comment]
```python
In [23]: a = np.zeros(2**30, 'H')

In [24]: %time [(a.astype('f'),None)[1] for _ in range(99)][:0]
CPU times: user 40.9 s, sys: 47.6 s, total: 1min 28s
Wall time: 1min 28s
Out[24]: []

In [25]: a = np.zeros(2**30, 'h')

In [26]: %time [(a.astype('f'),None)[1] for _ in range(99)][:0]
CPU times: user 17.9 s, sys: 46.6 s, total: 1min 4s
Wall time: 1min 4s
Out[26]: []

In [27]: a = torch.zeros(2**30, dtype=torch.int16)

In [28]: %time [(a.float(),None)[1] for _ in range(99)][:0]
CPU times: user 1min 27s, sys: 2min 32s, total: 4min
Wall time: 50 s
Out[28]: []
```

A more detailed test. `numpy`'s version is definitely around 4 times faster than `pytorch`'s version. I did this twice and yielded the same result. Not sure about the `2min 32s` sys time of `pytorch`, considering they are processing the same amount of data, and `sys` time usually means cache miss?

**Comment by vadimkantorov** [comment]
> I'm not able to reproduce your 1000x number, but I do sometimes see spikes. (It seems to be transient, and generally not more than 2x.) Maybe grab some perf stats and see what it's blocking on?

It’s definitely multi-threading influence on time.time, since real time difference is much less. Any advice on how to collecting stats about this? Should I call gettimeofday manually somehow?

**Comment by vadimkantorov** [comment]
@robieta If you have advice on how to collect metrics with perf tools, I'll do that as well

**Comment by robieta** [comment]
@ZisIsNotZis The relevant measurement here is `Wall time`, where you can see PyTorch is on par with NumPy.

@vadimkantorov Generally start with tutorials on Linux's `perf` tool.

**Comment by vadimkantorov** [comment]
I've tried to use `perf` tool e.g. for collecting traces of ffmpeg and found that it's super-important to have library built with debug symbols, otherwise stack traces labels are anonymous. Is there a provided PyTorch image with debug symbols?

**Comment by robieta** [comment]
You'll have to build it yourself:

https://github.com/pytorch/pytorch/blob/master/CONTRIBUTING.md#build-only-what-you-need

Use `REL_WITH_DEB_INFO`, as `DEBUG` will build without optimizations.

**Comment by vadimkantorov** [comment]
thanks for the pointer! will take a look when again will get back to optimizing that code path

**Comment by vadimkantorov** [comment]
It would be cool to have some public benchmarks on pytorch.org comparing PyTorch versus NumPy for some popular CPU workloads

**Comment by vadimkantorov** [comment]
This fused multiplication + conversion should be a good test case for fuser / fx / or whatever the latest pytorch compiler technology is called :)

**Comment by CaoE** [comment]
I have similar observations as ZisIsNotZis. 
When the problem size is not big, numpy is better than pytorch (mulit-thread). As problem size increase, the performance of 
pytorch become better：
```
1. problem size = 10000:
torch: 2.770662307739258 us
numpy: 1.6150474548339844 u

2. problem size = 40000:
torch: 12.48478889465332 us
numpy: 4.920005798339844 us

3. problem size = 100000:
torch: 12.502193450927734 us
numpy: 11.224508285522461 us

4. problem size = 400000:
torch: 13.741731643676758 us
numpy: 23.760318756103516 us
```

Pytorch does not leverage vectorized conversion for int16->fp32 since it is not supported yet.
As robieta mentioned above,  NumPy seems has specialized implementations and it is faster than pytorch just considering the overhead of the conversion of int16 -> fp32 that's why when problem size is not big numpy is better than pytorch.
When problem size become larger, L2 or L3 cache miss ratio become higher and the overhead of data load and save takes higher ratio.  Finally, torch seems have better parallel mechanism.
So in summary, the performance of the pure conversion of int16->fp32 in torch is slower (Need to add specialization for int6->fp32 and add vectorized conversion method for int16->fp32), but as the scale of the problem increases, it will be alleviated.

@vadimkantorov If it is necessary to improve the performance of int16->fp32 in certain cases, perhaps you can request a feature from pytorch.

**Comment by vadimkantorov** [comment]
thanks for the detailed investigation :)
renamed this issue
